src/mt-system-channel.js

```diff
@@ -0,0 +1,234 @@
+  }
+
+  function update_connected(system_name, ws_cx) {
+    if (!connection_bus[system_name]) {
+      const sys_duplex = new PassThrough();
+
+      connection_bus[system_name] = { ws_cx, q: sys_duplex };
+
+      sys_duplex.on('data', function(chunk) {
+        const my_cx = connection_bus[system_name].ws_cx;
+
+        if (my_cx && my_cx.connected) {
+          my_cx.send(chunk.toString());
+        } else {
+          sys_duplex.pause();
+          sys_duplex.unshift(chunk);
+        }
+      });
+
+    } else if (ws_cx) {
+      connection_bus[system_name].ws_cx = ws_cx;
+
+      if (waiting_connections[system_name]) {
+        waiting_connections[system_name]();
+      }
+
+      if (connection_bus[system_name].q.isPaused()) {
+        connection_bus[system_name].q.resume();
+      }
+    }
+  }
+
+  /**
+   * This method allows the user to call a one-time function when the indicated system connects
+   * to the gateway.
+   * @param {String} system_name The system we're waiting to hear is connected
+   * @param {Function} cb One-time function to execute when this system connects
+   */
+  function await_connection(system_name, cb) {
+    if (!system_name || typeof system_name !== 'string') {
+      throw new Error(
+        'await_connection takes a system name string as its first argument'
+      );
+    }
+
+    if (typeof cb !== 'function') {
+      throw new Error(
+        'await_connection requires a callback function as its second argument'
+      );
+    }
+
+    waiting_connections[system_name] = cb;
+  }
+
+  function on_http_request(cb) {
+    if (typeof cb !== 'function') {
+      throw new Error('on_http_request requires a function as its argument');
+    }
+
+    http_request_callback = cb;
+  }
+
+  function system_message_has_errors(message) {
+    // TODO: Validate the passed message against the system websocket interface api.
+    return false;
+  }
+
+  ws_server.on('request', function (incoming) {
+    const { origin, resource } = incoming;
+    const system_name = valid_resource_request(resource);
+
+    if (!system_name) {
+      incoming.reject();
+      post_message(
+        // TODO: This probably should be an event or warning to MT? A system attempted to connect
+        // but was rejected by the gateway...?
+        `${new Date()} : Connection from origin ${origin} rejected`
+      );
+    } else {
+      const connection = incoming.accept(null, origin);
+
+      connection.on('message', function(ws_message) {
+        const { data, utf8Data } = ws_message;
+        const message = data || utf8Data;
+
+        return (
+          // returning a value from an individual system callback or from any_system_callback will
+          // prevent the lib from automatically sending the message to MT.
+          (system_callbacks[system_name] && system_callbacks[system_name](message, system_name)) ||
+          (any_system_callback && any_system_callback(message)) ||
+          post_message(message)
+        );
```

Comment:
this logic here has gotten a bit out of hand. are you trying to avoid side-effects by not returning `false` to the EventEmitter / websocket library? after reading NodeJS and WebSocket docs I couldn't figure out why that would be the case.. I think this could be a lot simpler by just early returning in each case, i.e...

```suggestion
        if (system_callbacks[system_name]) {
          system_callbacks[system_name](message, system_name);
          return;
        }
        
        if (any_system_callback) {
          any_system_callback(message);
          return;
        }
        
        post_message(message);
```